### PR TITLE
Async Metrics

### PR DESCRIPTION
**Pull request by felixdesouza**
**Goals (and why)**:
As part of some of the conjure-undertow changes, async versions of requests were added. The sad part is, the metrics as we know currently are off since they measure the time it takes to be given back the `ListenableFuture` vs how long the `ListenableFuture` takes to complete.

Ideally this would go in https://github.com/palantir/tritium, but given that we've hacked together the `TaggedMetricsInvocationHandler` for our own (nefarious?) gains, I didn't want to unwind that change as I'm working on Timelock Partitioning etc. @schlosna for SA. My proposed impl would be that it goes in `ByteBuddyInstrumentationAdvice` or `InvocationEventProxy` then every `InvocationHandler` gets async metrics stuff for free!

**Implementation Description (bullets)**:
* Backport some of the metrics changes in the `TaggedMetricsInvocationHandler` wrt metric name cache etc. Since the metric name doesn't change for most invocations we're wasting time doing the string concat and the allocations for `MetricName`. For the ones with the extra tags, have a specialised cache for those calls.
* Check if the return type is `ListenableFuture` and if so, add a `listener` that will update the metric. Otherwise we leave it as is.

**Testing (What was existing testing like?  What have you done to improve it?)**:
* Hard to test properly since we can't change the future that gets returned, we can only act on it => we can have a race condition where if we were to wait for the future to be done, it's not necessarily the case that the metrics would have been tracked correctly. Only fix I see for this is in upstream in tritium.

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:
* `TaggedMetricsEventInvocationHandler`
* `AtlasDbMetrics`

**Priority (whenever / two weeks / yesterday)**:
ASAP! Some metrics in our internal dashboard is a lie until fixed.

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java [hunk 1]
```diff
@@ -71,18 +71,27 @@ private AtlasDbMetrics() {}
...
+            Class<T> serviceInterface,
+            U service) {
+        return Instrumentation.builder(serviceInterface, service)
+                .withHandler(new TaggedMetricsInvocationEventHandler(
+                        taggedMetrics,
+                        MetricRegistry.name(serviceInterface)))
+                .withPerformanceTraceLogging()
+                .build();
+    }
+
```
Comment (Jolyon-S) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: This implementation is very nearly identical to the one below - is there a common method that can be extracted?

Also, below you factored out `String name = MetricRegistry.name(serviceInterface);`, but not in this block.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java [hunk 2]
```diff
@@ -48,7 +58,26 @@
...
+    private final ConcurrentMap<Method, Timer> untaggedTimerCache = Maps.newConcurrentMap();
+
+    @Nullable
+    private final Function<MethodWithExtraTags, Timer> onSuccessTimerMappingFunctionExtraTags;
+    private final ConcurrentMap<MethodWithExtraTags, Timer> extraTagsTimerCache = Maps.newConcurrentMap();
+
+    public TaggedMetricsInvocationEventHandler(TaggedMetricRegistry taggedMetricRegistry, String serviceName) {
+        super(InstrumentationUtils.getEnabledSupplier(serviceName));
+        this.taggedMetricRegistry = Preconditions.checkNotNull(taggedMetricRegistry, "metricRegistry");
+        this.serviceName = Preconditions.checkNotNull(serviceName, "serviceName");
```
Comment (Jolyon-S) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: Repeated code from below; is there a better way this can be done?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java [hunk 3]
```diff
@@ -72,12 +106,54 @@ public void onSuccess(@Nullable InvocationContext context, @Nullable Object resu
             return;
         }
 
+        if (result != null && ListenableFuture.class.isAssignableFrom(context.getMethod().getReturnType())) {
+            Futures.addCallback((ListenableFuture<?>) result, new FutureCallback<Object>() {
```
Comment (Jolyon-S) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: It looks like you are casting `result` from `Object` to `ListenableFuture<?>`; can you do this without a check? (It looks to me like the check above is checking `context`'s return type, but I don't know if that is guaranteed to be the same as the type of `result`).

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java [hunk 4]
```diff
@@ -95,7 +171,6 @@ public void onFailure(@Nullable InvocationContext context, @Nonnull Throwable ca
         taggedMetricRegistry.meter(MetricName.builder().safeName(
                 MetricRegistry.name(failuresMetricName, cause.getClass().getName())).build())
                 .mark();
-
```
Comment (Jolyon-S) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: nit: unnecessary diff?

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java [hunk 5]
```diff
@@ -91,6 +122,77 @@ public void instrumentWithCustomNameAll() {
...
+            // transformed future
+            Awaitility.await()
+                    .atMost(ASYNC_DURATION_TTL.toMillis(), TimeUnit.MILLISECONDS)
+                    .until(() -> taggedMetrics.timer(metricName).getSnapshot().size() > 0);
+            return true;
+        }, MoreExecutors.directExecutor());
+
+        assertThat(done.get()).isEqualTo(true);
+
+        assertThat(done.get()).isEqualTo(true);
```
Comment (Jolyon-S) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: Should there be two of the exact same statement here? If so, please may you leave a comment as to why?

nit: maybe also group them together w.r.t spacing?

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java [hunk 6]
```diff
@@ -91,6 +122,77 @@ public void instrumentWithCustomNameAll() {
...
+                .as("in the event of scheduling issues, and despite having 10 concurrent futures, we complete "
+                        + "everything with 2*ttl")
+                .isBefore(now.plus(ASYNC_DURATION_TTL.plus(ASYNC_DURATION_TTL)));
+        Snapshot snapshot = taggedMetrics.timer(metricName).getSnapshot();
+        assertThat(Duration.ofNanos(snapshot.getMin())).isGreaterThan(ASYNC_DURATION_TTL);
+        assertThat(snapshot.size()).isEqualTo(10);
+    }
+
+    @Test
+    public void instrumentTaggedAsyncFunctionWithExtraTags() throws InterruptedException, ExecutionException {
```
Comment (Jolyon-S) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: Copied code from above - can a common method be factored out?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java (hunk 1 above)
Comment (felixdesouza) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: can do, not really sure it buys anything though, this is what it looks like:

```java
    public static <T, U extends T> T instrumentWithTaggedMetrics(
            TaggedMetricRegistry taggedMetrics,
            Class<T> serviceInterface,
            U service) {
        return buildInstrumentation(
                serviceInterface,
                service,
                new TaggedMetricsInvocationEventHandler(taggedMetrics, MetricRegistry.name(serviceInterface)));
    }

    public static <T, U extends T> T instrumentWithTaggedMetrics(
            TaggedMetricRegistry taggedMetrics,
            Class<T> serviceInterface,
            U service,
            Function<InvocationContext, Map<String, String>> tagFunction) {
        return buildInstrumentation(
                serviceInterface,
                service,
                new TaggedMetricsInvocationEventHandler(
                        taggedMetrics,
                        MetricRegistry.name(serviceInterface),
                        tagFunction));
    }

    private static <T, U extends T> T buildInstrumentation(
            Class<T> serviceInterface,
            U service,
            TaggedMetricsInvocationEventHandler handler) {
        return Instrumentation.builder(serviceInterface, service)
                .withHandler(handler)
                .withPerformanceTraceLogging()
                .build();
    }
```

The factoring was so it would not kill my line count aha.

Comment (Jolyon-S) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/util/AtlasDbMetrics.java]: Yeah this isn't particularly pretty. Don't mind either way, so long as you are consistent (i.e. the factoring thing is either done or not done for both.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java (hunk 5 above)
Comment (felixdesouza) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: deleted, can't remember, was likely an accident

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java (hunk 6 above)
Comment (felixdesouza) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: some code can be factored out, I won't go the full length because it's important to see what's actually in the test without having to jump around everywhere.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java (hunk 3 above)
Comment (felixdesouza) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: they should be equivalent, I suppose I wanted to adhere to the interface in question as opposed to whatever type we may get back, I can add another check to the object itself.

File: atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java (hunk 6 above)
Comment (Jolyon-S) [reply atlasdb-client/src/test/java/com/palantir/atlasdb/util/AtlasDbMetricsTest.java]: That's fair, it just seemed that a fair amount was exact copy/paste (but I don't mind things with small variations not being factored out exactly).
